src/buffer_manager.ts

```diff
@@ -622,8 +624,39 @@ export class BufferManager implements Disposable {
             this.textDocumentToBufferId.delete(document);
             unusedBuffers.push(bufId);
         });
-        unusedWindows.length && (await actions.lua("close_windows", unusedWindows));
-        unusedBuffers.length && (await actions.lua("delete_buffers", unusedBuffers));
+
+        logger.debug(`Closing windows: ${inspect(unusedWindows)}...`);
+        for (const winId of unusedWindows) {
+            logger.debug(`Closing window ${winId}...`);
+            try {
+                await this.request("nvim_win_close", [winId, true], 5000);
+            } catch (err) {
+                // TODO: should we avoid deleting from editor<->window map if it failed to close here?
+                logger.warn(err);
+            }
+        }
+
+        logger.debug(`Deleting buffers: ${inspect(unusedBuffers)}...`);
+        for (const buf of unusedBuffers) {
+            logger.debug(`Deleting buffer ${buf}...`);
+            try {
+                await this.request("nvim_buf_delete", [buf, { force: true }], 5000);
+            } catch (err) {
+                // TODO: same question for document<->buffer map
+                logger.warn(err);
+            }
+        }
+    }
+
+    // Workaround for #2136 - this request sometimes hangs indefinitely so we
+    // just hardcode it to return after some time, to restore functionality to the user.
+    private async request(name: string, args: any[], delayMs: number): Promise<void> {
+        return Promise.race([
+            this.client.request(name, args),
+            wait(delayMs).then(() => {
```

Comment:
Hmm. Maybe the [client](https://github.com/neovim/node-client/) `request()` should have a builtin way to support a "timeout" parameter.

I guess there's no way to cancel the request :/